File: Examples/Simulation_C/Simulation_3C.m
% Multivariate Bayesian Inversion for Classification and Regression
% Simulation 3: comparison of methods (MATLAB script)
% 
% Author: Joram Soch, OvGU Magdeburg
% E-Mail: joram.soch@ovgu.de
% 
% Version History:
% - 20/02/2022, 15:55: first MATLAB version
% - 20/02/2025, 14:57: final MATLAB version
% - 26/02/2026, 10:57: added performance plot
% - 26/02/2026, 11:08: rewrote MBC and SVC
% - 26/02/2026, 11:21: added LDA classification
% - 26/02/2026, 11:32: added GNB classification
% - 26/02/2026, 11:43: added LogReg classification
% - 26/02/2026, 12:01: added RF classification
% - 26/02/2026, 12:15: added NN classification
% - 26/02/2026, 12:44: refined performance plot


clear
close all

%%% Step 1: specify ground truth %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% set ground truth
rng(1);
mu = 1;                         % class means
Si = [1, 0.5; 0.5, 1];          % covariance structure
s2 = 4;                         % noise variance
n  = 300;                       % number of data points
k  = 10;                        % number of CV folds
v  = 2;                         % number of features
C  = 3;                         % number of classes

% generate classes
x  = [kron([1:C]',ones(n/C,1)), rand(n,1)];
x  = sortrows(x,2);
x  = x(:,1);                    % randomized labels
X  = zeros(n,C);                % design matrix
V  = eye(n);                    % observation covariance
for i = 1:n
    X(i,x(i)) = 1;
end;


%%% Step 2: generate data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% generate data
B = [  -mu,   +mu;
     +2*mu, +2*mu;
       +mu,   -mu];
E = matnrnd(zeros(n,v), s2*V, Si, 1);
Y = X*B + E;


%%% Step 3: analyze data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% specify analysis parameters
meth    = {'MBC', 'GNB', 'LDA', 'LogReg', 'SVC', 'RFC', 'NNC'};
prior.x = [1:C];                % MBC class indices
prior.p = 1/C*ones(1,C);        % MBC prior probabilities
Dgnb    = 'normal';             % GNB distribution name
Dlda    = 'linear';             % LDA discriminant type
Llogreg = 'logit';              % LogReg link function
Mlogreg = 'nominal';            % LogReg model type
Csvm    = 1;                    % SVM cost parameter
Mrf     = 'Bag';                % RF aggregation method
Lnn     = [10, 10];             % NN hidden layer sizes
Ann     = 'sigmoid';            % NN activation function
CV      = ML_CV(x, k, 'kfc');   % k-fold CV on points per class

% preallocate results
M  = numel(meth);
xp = zeros(n,M);
CA = zeros(1,M);

% perform classification
for g = 1:size(CV,2)

    % get training and test set
    i1 = find(CV(:,g)==1);
    i2 = find(CV(:,g)==2);
    Y1 = Y(i1,:);
    Y2 = Y(i2,:);
    x1 = x(i1);
    x2 = x(i2);
    V1 = V(i1,i1);
    V2 = V(i2,i2);

    % evaluate all methods
    for h = 1:M
        
        % Method: multivariate Bayesian classification (MBI)
        % https://github.com/JoramSoch/MBI/blob/main/MATLAB/ML_MBI.m
        if strcmp(meth{h}, 'MBC')
            mba1 = mbitrain(Y1, x1, [], V1, 'MBC');
            pp2  = mbitest(Y2, x2, [], V2, mba1, prior);
            [m,j]= max(pp2, [], 2);
            xp2  = prior.x(j);
        end;
        
        % Method: Gaussian naive Bayes (MATLAB)
        % https://de.mathworks.com/help/stats/naive-bayes-classification.html
        if strcmp(meth{h}, 'GNB')
            nbc1 = fitcnb(Y1, x1, 'DistributionNames', Dgnb, 'Prior', prior.p);
            xp2  = predict(nbc1, Y2);
        end;
        
        % Method: linear discriminant analysis (MATLAB)
        % https://de.mathworks.com/help/stats/discriminant-analysis.html
        if strcmp(meth{h}, 'LDA')
            lda1 = fitcdiscr(Y1, x1, 'DiscrimType', Dlda);
            xp2  = predict(lda1, Y2);
        end;
        
        % Method: multinomial logistic regression (MATLAB)
        % https://de.mathworks.com/help/stats/fitmnr.html
        if strcmp(meth{h}, 'LogReg')
            logreg1 = fitmnr(Y1, x1, 'Link', Llogreg, 'ModelType', Mlogreg);
            xp2     = predict(logreg1, Y2);
        end;
        
        % Method: support vector classification (LibSVM)
        % https://github.com/JoramSoch/ML4ML/blob/main/ML_SVC.m
        if strcmp(meth{h}, 'SVC')
            opt  = sprintf('-s 0 -t 0 -c %s -q', num2str(Csvm));
            svm1 = svmtrain(x1, Y1, opt);
            xp2  = svmpredict(x2, Y2, svm1, '-q');
        end;

        % Method: random forrest classification (MATLAB)
        % https://de.mathworks.com/help/stats/select-predictors-for-random-forests.html
        if strcmp(meth{h}, 'RFC')
            rfe1 = fitcensemble(Y1, x1, 'Method', Mrf);
            xp2  = round(predict(rfe1, Y2));
        end;

        % Method: neural network classification (MATLAB)
        % https://de.mathworks.com/help/stats/classificationneuralnetwork.html
        if strcmp(meth{h}, 'NNC')
            nn1 = fitcnet(Y1, x1, 'LayerSizes', Lnn, 'Activations', Ann);
            xp2 = predict(nn1, Y2);
        end;
        
        % store test set predictions
        xp(i2,h) = xp2;
        
    end;
    clear mba1 nbc1 lda1 logreg1 svm1 rfe1 nn1
    
end;

% calculate performance
for h = 1:M
    CA(h) = mean(xp(:,h)==x);
end;


%%% Step 4: visualize results %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% open figure
figure('Name', 'Simulation 3 (Comparison)', 'Color', [1 1 1], 'Position', [50 50 800 900]);
cols  = 'bcgmryw';
x_off = 3;
y_off = 0.9;
hold on;

% plot performance
for h = 1:M
    bar(h, CA(h), 0.7, cols(h));
end;
plot([(1-1), (M+1)], [1/C, 1/C], ':k', 'LineWidth', 2);
plot([x_off, x_off]+1/2, [0, 1], '-k', 'LineWidth', 1);
axis([(1-1), (M+1), 0, 1]);
set(gca,'Box','On');
set(gca,'XTick',[1:M],'XTickLabel',meth);
legend([meth, {'chance'}], 'Location', 'NorthEast');
xlabel('classification approach', 'FontSize', 16);
ylabel('classification accuracy', 'FontSize', 16);
title('Simulation 3: Comparison', 'FontSize', 24);
text(x_off+1/2, y_off, sprintf('generative   \nmethods   '), ...
     'FontSize', 16, 'HorizontalAlignment', 'right', 'VerticalAlignment', 'middle');
text(x_off+1/2, y_off, sprintf('   discriminative\n   methods'), ...
     'FontSize', 16, 'HorizontalAlignment', 'left', 'VerticalAlignment', 'middle');

File: Examples/Simulation_C/Simulation_3C.py
# -*- coding: utf-8 -*-
"""
Multivariate Bayesian Inversion for Classification and Regression
Simulation 3: comparison of methods (Python script)

Author: Joram Soch, OvGU Magdeburg
E-Mail: joram.soch@ovgu.de

Version History:
- 2022-02-20, 12:23: first version in MATLAB
- 2025-03-05, 10:58: final version in Python
- 2026-06-16, 19:49: rewrote MBC and SVC,
                     implemented RGA measure
- 2026-06-17, 12:38: added GNB, LDA, LogReg
- 2026-06-17, 14:27: added RFC, NNC
"""


# specify MBI path
MBI_dir = '../../Python/'

# import packages
import os
import numpy as np
import sklearn as skl
import matplotlib.pyplot as plt

# import MBI module
orig_dir = os.getcwd()
os.chdir(MBI_dir)
import MBI
os.chdir(orig_dir)


### Step 1: specify ground truth ##############################################

# set ground truth
np.random.seed(2)
mu = 1                                      # class means
Si = np.array([[1,0.5],[0.5,1]])            # feature covariance
s2 = 4                                      # noise variance
n  = 300                                    # number of data points
k  = 10                                     # number of CV folds
v  = 2                                      # number of features
C  = 3                                      # number of classes

# generate classes
x  = np.kron(np.arange(C).reshape((C,1))+1, np.ones((int(n/C),1)))
x  = np.random.permutation(x[:,0])          # randomized labels
X  = np.zeros((n,C))                        # design matrix
V  = np.eye(n)                              # observation covariance
for i in range(n):
    X[i,int(x[i]-1)] = 1


### Step 2: generate data #####################################################

# generate data
B = np.array([[  -mu,   +mu],
              [+2*mu, +2*mu],
              [  +mu,   -mu]])
E = MBI.matnrnd(np.zeros((n,v)), s2*V, Si, 1)
Y = X @ B + E


### Step 3: analyze data ######################################################

# specify analysis parameters
meth    = ['GNB', 'MBC', 'LDA', 'LogReg', 'SVC', 'RFC', 'NNC']
prior   = {'x': np.arange(C)+1,             # MBC class indices
           'p': 1/C*np.ones(C)}             # MBC prior probabilities
Dgnb    = 'normal'                          # GNB distribution name
Dlda    = 'linear'                          # LDA discriminant type
Llogreg =  1                                # LogReg cost parameter
Ksvm    = 'linear'                          # SVM kernel type
Csvm    =  1                                # SVM cost parameter
Nrf     = 100                               # RF number of trees
Lnn     = (10, 10)                          # NN hidden layer sizes
Ann     = 'logistic'                        # NN activation function

# specify cross-validation
MBC = MBI.cvMBI(Y, x, V=V, mb_type='MBC')
MBC.crossval(k=k, cv_mode='kfc')            # k-fold CV on points per class
CV  = MBC.CV

# preallocate results
M   = len(meth)
xp  = np.zeros((n,M))
xs  = np.zeros((n,C,M))
CA  = np.zeros(M)
RGA = np.zeros(M)

# perform classification
for g in range(k):

    # get training and test set
    i1 = np.array(np.nonzero(MBC.CV[:,g]==1)[0], dtype=int)
    i2 = np.array(np.nonzero(MBC.CV[:,g]==2)[0], dtype=int)
    Y1 = Y[i1,:]
    Y2 = Y[i2,:]
    x1 = x[i1]
    x2 = x[i2]
    V1 = V[i1,:][:,i1]
    V2 = V[i2,:][:,i2]
    
    # evaluate all methods
    for h in range(M):
        
        # prepare test set predictions
        xp2  = np.zeros(i2.size)
        xs2  = 1/C*np.ones((i2.size,C))
        
        # Method: multivariate Bayesian classification (MBI)
        # https://github.com/JoramSoch/MBI/blob/main/Python/MBI.py
        if meth[h] == 'MBC':
            mba1 = MBI.model(Y1, x1, V=V1, mb_type='MBC').train()
            pp2  = MBI.model(Y2, x2, V=V2, mb_type='MBC').test(mba1, prior)
            xp2  = np.argmax(pp2, axis=1) + 1
            xs2  = pp2
        
        # Method: Gaussian naive Bayes (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html
        if meth[h] == 'GNB':
            gnb1 = skl.naive_bayes.GaussianNB().fit(Y1, x1)
            xp2  = gnb1.predict(Y2)
            xs2  = gnb1.predict_proba(Y2)

        # Method: linear discriminant analysis (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html
        if meth[h] == 'LDA':
            lda1 = skl.discriminant_analysis.LinearDiscriminantAnalysis().fit(Y1, x1)
            xp2  = lda1.predict(Y2)
            xs2  = lda1.predict_proba(Y2)

        # Method: logistic regression (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
        if meth[h] == 'LogReg':
            logreg1 = skl.linear_model.LogisticRegression(C=Llogreg).fit(Y1, x1)
            xp2     = logreg1.predict(Y2)
            xs2     = logreg1.decision_function(Y2)

        # Method: support vector classification (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html
        if meth[h] == 'SVC':
            svm1 = skl.svm.SVC(kernel=Ksvm, C=Csvm).fit(Y1, x1)
            xp2  = svm1.predict(Y2)
            xs2  = svm1.decision_function(Y2)
        
        # Method: random forrest classification (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
        if meth[h] == 'RFC':
            rf1 = skl.ensemble.RandomForestClassifier(
                n_estimators=Nrf).fit(Y1, x1)
            xp2 = rf1.predict(Y2)
            xs2 = rf1.predict_proba(Y2)
        
        # Method: neural network classification (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html
        if meth[h] == 'NNC':
            nn1 = skl.neural_network.MLPClassifier(
                hidden_layer_sizes=Lnn, activation=Ann).fit(Y1, x1)
            xp2 = nn1.predict(Y2)
            xs2 = nn1.predict_proba(Y2)
        
        # store test set predictions
        xp[i2,h]   = xp2
        xs[i2,:,h] = xs2
        
    # delete analysis variables  
    del mba1, pp2, gnb1, lda1, svm1

# calculate performance
for h in range(M):
    CA[h]  = np.mean(xp[:,h]==x)
    RGA[h] = MBI.calc_RGA(x, xs[:,:,h], 'macro')


### Step 4: visualize results #################################################

# open figure
fig   = plt.figure(figsize=(8,9))
ax    = fig.add_subplot(111)
x_off = 3
y_off = 0.9
cols  = [[  0,  32,  96], [  0,   0, 255], [  0, 176, 240],
         [192,   0,   0], [255,   0,   0], [255, 192,   0], [255, 255,   0]]

# plot performance
for h in range(M):
    col = tuple([rgb/255 for rgb in cols[h]])
    ax.bar(h+1, RGA[h], width= 0.7, color=col, edgecolor='k', label=meth[h])
ax.plot([(1-1), (M+1)], [1/2, 1/2],     ':k', linewidth=2, label='chance')
ax.plot([x_off+1/2, x_off+1/2], [0, 1], '-k', linewidth=1)
ax.axis([(1-1), (M+1), 0, 1])
ax.set_xticks([])
ax.legend(loc='lower right', fontsize=12)
ax.set_xlabel('classification approach', fontsize=20)
ax.set_ylabel('rank graduation accuracy', fontsize=20)
ax.set_title('Simulation 3', fontsize=24, fontweight='bold')
ax.tick_params(axis='both', labelsize=12)
ax.text(x_off+1/2, y_off, 'generative   \nmethods   ',
        fontsize=20, ha='right', va='center')
ax.text(x_off+1/2, y_off, '   discriminative\n   methods',
        fontsize=20, ha='left', va='center')

# enable tight layout
fig.tight_layout()